### PR TITLE
Namespaced dataset types/dataset modules

### PR DESCRIPTION
**Pull request by bdmogal**
Have to re-open a PR for #1354 because I'm hitting some github bug when you use rebase with integration branches.

Description from original PR: 

Summary:

Going to do namespaced datasets in 3 PRs:
This PR: Namespaced dataset modules, types

* **Incompatible change**: ``DatasetTypeMDS`` extends ``MetadataStoreDataset`` and uses ``MDSKey`` as the key to keep it consistent with ``AppMetadataStore``. Similar change for ``DatasetInstanceMDS`` will come in a following PR.
* ``DatasetTypeHandler`` is namespaced. 
* Dataset type loading logic: Try to find in current namespace first. If not found, look for it in the system namespace.
* Upgrade tool for these changes is a WIP, will happen before Monday.

Two other PRs coming up, before merge to develop.
1. Updates to NamespacedDatasetFramework to use namespaces
2. Namespaced dataset instances.
Cannot merge to develop individually because some APIs cannot work without the complete integration.

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DeployDatasetModulesStage.java [hunk 1]
```diff
@@ -64,12 +64,12 @@ public void process(ApplicationDeployable input) throws Exception {
...
         // note: it seems dangerous to instantiate dataset module here, but this will be fine when we move deploy into
         //       isolated user's environment (e.g. separate yarn container)
-        Id.DatasetModule moduleId = Id.DatasetModule.from(input.getId().getNamespaceId(), moduleName);
+        Id.DatasetModule moduleId = Id.DatasetModule.from(input.getId().getNamespace(), moduleName);
         if (DatasetModule.class.isAssignableFrom(clazz)) {
           datasetFramework.addModule(moduleId, (DatasetModule) clazz.newInstance());
         } else if (Dataset.class.isAssignableFrom(clazz)) {
           // checking if type is in already
-          if (!datasetFramework.hasType(clazz.getName())) {
+          if (!datasetFramework.hasSystemType(clazz.getName())) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DeployDatasetModulesStage.java]: don't we still need to check that it doesn't already have that type in its namespace in addition to checking the system?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceHandler.java [hunk 2]
```diff
@@ -303,7 +337,11 @@ private boolean dropDataset(Id.DatasetInstance datasetInstaceId, DatasetSpecific
       return false;
     }
 
-    opExecutorClient.drop(datasetInstaceId, implManager.getTypeInfo(spec.getType()), spec);
+    DatasetTypeMeta typeMeta = getTypeInfo(datasetInstaceId.getNamespace(), spec.getType());
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceHandler.java]: not part of your PR, but 'datasetInstaceId' is not spelled correctly (should be datasetInstanceId)

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/mds/DatasetTypeMDS.java [hunk 3]
```diff
@@ -32,91 +36,117 @@
...
 
-  public DatasetTypeMeta getType(String typeName) {
-    DatasetModuleMeta moduleName = getModuleByType(typeName);
+  public DatasetTypeMeta getType(Id.DatasetType datasetTypeId) {
+    DatasetModuleMeta moduleName = getModuleByType(datasetTypeId);
     if (moduleName == null) {
       return null;
     }
-    return getTypeMeta(typeName, moduleName);
+    return getTypeMeta(datasetTypeId.getNamespace(), datasetTypeId.getTypeName(), moduleName);
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/mds/DatasetTypeMDS.java]: may be better to just pass the whole datasetTypeId

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/mds/DatasetTypeMDS.java [hunk 4]
```diff
@@ -32,91 +36,117 @@
...
-  private DatasetTypeMeta getTypeMeta(String typeName, DatasetModuleMeta moduleMeta) {
+  private DatasetTypeMeta getTypeMeta(Id.Namespace namespaceId, String typeName, DatasetModuleMeta moduleMeta) {
     List<DatasetModuleMeta> modulesToLoad = Lists.newArrayList();
     // adding first all modules we depend on, then myself
     for (String usedModule : moduleMeta.getUsesModules()) {
-      modulesToLoad.add(getModule(usedModule));
+      // Try to find module in the specified namespace first
+      DatasetModuleMeta usedModuleMeta = getModule(Id.DatasetModule.from(namespaceId, usedModule));
+      // if not found, try to load it from system namespace
+      if (usedModuleMeta == null) {
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/mds/DatasetTypeMDS.java]: should this fallback logic be in ```getModule``` or is there some reason we would not want to look in the system namespace for that method?

Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/mds/DatasetTypeMDS.java]: i guess it would be weird if it did since the id contains namespace

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/InMemoryDatasetFramework.java [hunk 5]
```diff
@@ -52,18 +53,21 @@
  */
 public class InMemoryDatasetFramework implements DatasetFramework {
   private static final Logger LOG = LoggerFactory.getLogger(InMemoryDatasetFramework.class);
+  private static final Id.Namespace systemNamespace = Id.Namespace.from(Constants.SYSTEM_NAMESPACE);
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/InMemoryDatasetFramework.java]: may be useful to have a ```Id.Namespace.SYSTEM``` constant instead of using the string everywhere.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/InMemoryDatasetFramework.java [hunk 6]
```diff
@@ -233,13 +256,57 @@ private DatasetDefinitionRegistry createRegistry(@Nullable ClassLoader classLoad
...
 
+  private List<String> getAvailableModuleClasses(Id.Namespace namespace) {
+    LinkedList<String> allClasses = getAllModuleClasses(systemNamespace);
+    if (systemNamespace.equals(namespace)) {
+      return allClasses;
+    }
+    LinkedList<String> namespaceClasses = getAllModuleClasses(namespace);
+    for (String namespaceClass : namespaceClasses) {
+      int idx = allClasses.indexOf(namespaceClass);
+      if (idx > -1) {
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/InMemoryDatasetFramework.java]: im confused about what this is doing, we're removing then adding it right back?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DeployDatasetModulesStage.java (hunk 1 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DeployDatasetModulesStage.java]: Yeah, I wanted to get feedback on this. The current logic is that we will not upgrade if a module exists, irrespective of whether its a user module. Ideally we should allow upgrading user modules by default, but I understand there are some limitations regarding this in CDAP currently.

I'm unsure which way to go here amongst the following 3 options:
1. Allow upgrade no matter what - even if it is a module that exists in the system namespace, we create a copy of it in another namespace and allow users to use it. 
2. Allow upgrade only for user modules - the logic that is used in this PR.
3. Do not allow upgrade at all - the current logic.

I guess all three are possible from the namespaces point of view, however, I understand that other things in datasets can break with upgrades.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/InMemoryDatasetFramework.java [hunk 7]
```diff
@@ -277,7 +357,19 @@ public void add(DatasetDefinition def) {
 
     @Override
     public <T extends DatasetDefinition> T get(String datasetTypeName) {
-      return delegate.get(datasetTypeName);
+      // In real-world scenarios, default modules are guaranteed to always exist in the system namespace.
+      // Hence, we could add a preconditions check here to verify that registries contains types from system namespace
+      // However, a lot of our tests (AbstractDatasetTest) start without default modules, so not adding that check.
+      // In any case, the pattern here of first looking for the definition in own namespace, then in system is valid
+      // and the else block should throw an exception if the dataset type is not found in the current or the system
+      // namespace.
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/InMemoryDatasetFramework.java]: good comment

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceHandler.java (hunk 2 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceHandler.java]: I'll fix it.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/mds/DatasetTypeMDS.java (hunk 3 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/mds/DatasetTypeMDS.java]: That seems right, I'll check again

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/mds/DatasetTypeMDS.java (hunk 4 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/mds/DatasetTypeMDS.java]: Thought about that, then felt it was better to leave that to clients? Not sure if we would always need it. I'll think some more.

**Comment by albertshau** [comment]
couple comments otherwise looks good

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/InMemoryDatasetFramework.java (hunk 6 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/InMemoryDatasetFramework.java]: This is applies when you have the same type class in the system namespace as well as the current module's namespace. In such a scenario, when you reach here, ``allClasses`` has type classes from modules in the system namespace. While iterating over the type classes from modules in the current namespace, if you encounter a class that is already present in ``allClasses``, you remove it, then add the corresponding version of the class from the module in the current namespace. This is to give type classes in the current namespace precedence over type classes in the system namespace.

Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/InMemoryDatasetFramework.java]: Probably deserves a comment. Will add this explanation.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/InMemoryDatasetFramework.java (hunk 5 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/InMemoryDatasetFramework.java]: Added ``Constants.SYSTEM_NAMESPACE_ID``

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/mds/DatasetTypeMDS.java (hunk 4 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/mds/DatasetTypeMDS.java]: ``getModule`` is used in other places where you do not want to fallback - e.g. in ``deleteModule``, or when you want to get the details of a particular module in a particular namespace. 
I think I'll add another method in this class ``getModuleWithFallback()`` that does fallback when clients want it to happen.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DeployDatasetModulesStage.java (hunk 1 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DeployDatasetModulesStage.java]: Going back to option 3. above. Will change it in a later PR if the need arises.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceHandler.java (hunk 2 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceHandler.java]: Fixed.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/mds/DatasetTypeMDS.java (hunk 3 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/mds/DatasetTypeMDS.java]: The ``getTypeMeta`` method does not always accept an ``Id.DatasetType``. It can accept ``(Id.Namespace, String typeName, DatasetModuleMeta)`` or ``(Id.Namespace, String typeName, Id.DatasetModule)``. Also, if you see the usage pattern else where, the namespace id and the typename are found separately (``getTypes`` method). So if I change the signature to accept an ``Id.DatasetType``, it would mean I would have to construct the ``Id.DatasetType`` when I have the namespace and the type name separately just for calling this method. 

Also, I feel that this signature calls out that the namespace belongs to both the module and the type.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DeployDatasetModulesStage.java (hunk 1 above)
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DeployDatasetModulesStage.java]: agreed, if we change it should be done in a separate pr

**Comment by albertshau** [comment]
one comment, then lgtm

**Comment by albertshau** [comment]
:+1:

**Comment by bdmogal** [comment]
For moving constants to ``Id.Namespace``, created - https://issues.cask.co/browse/CDAP-1554

**Comment by bdmogal** [comment]
Will wait for build and merge.

**Comment by bdmogal** [comment]
Build is green. merging.
